### PR TITLE
Integrate Authentication for Microsoft Accounts

### PR DESCRIPTION
**Pull request by Kashalls**
This pull request adds two new npm modules: `@xboxreplay/xboxlive-auth` and `node-fetch`
This integrates logging into Minecraft Servers for Microsoft Accounts. 
In order to log into a server using a microsoft account, you must specify options.auth = 'microsoft';

This fixes: 
https://github.com/PrismarineJS/node-minecraft-protocol/issues/782
https://github.com/PrismarineJS/mineflayer/issues/1480

// todo: clean up code just a little bit more.

### CODE REVIEW

File: src/client/encrypt.js [hunk 1]
```diff
@@ -18,7 +18,7 @@ module.exports = function (client, options) {
         client.end()
         return
       }
-      if (options.haveCredentials) {
+      if ((options.auth === 'microsoft' && !options.haveCredentials) || options.haveCredentials) {
```
Comment (Karang) [inline src/client/encrypt.js:21]: If you set `options.haveCredentials = true` in microsoftAuth.js, you don't need this change, and the PR will be a bit more self-contained.

Comment (Kashalls) [reply src/client/encrypt.js:21]: Well rom said he wanted to switch between the auth endpoints using auth and options.haveCredentials is a Mojang auth thing and auth is a Microsoft thing but if you want me to ditch the auth check here I can.

Comment (Karang) [reply src/client/encrypt.js:21]: `haveCredentials` is not a mojang thing https://github.com/PrismarineJS/node-minecraft-protocol/blob/master/src/client/auth.js#L22
its just a boolean to say if the user provided a password. If you are using a ms account it should have the same definition.

**Review by Karang — APPROVED** [review]
LGTM

**Comment by Kashalls** [comment]
> Does it work ? Does it fake the browser ?
> Can you update API.md with information on how to use it ?

It works, valid session and everything. Any requests to minecraftservices.com uses a simple user-agent, requests made to Xbox Live uses this User-Agent: `Mozilla/5.0 (XboxReplay; XboxLiveAuth/3.0) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/71.0.3578.98 Safari/537.36`

2FA is not supported with this method as it causes authentication issues, however this can be used to temporarily support it until a better solution is found.

**Comment by Kashalls** [comment]
> Can you update API.md with information on how to use it ?
I can, do you want me to create an example on how to use it with Microsoft Authentication?

**Comment by rom1504** [comment]
Yes, basically add the auth option in API.md, and a simple example is useful

**Comment by rom1504** [comment]
Thanks, that's great.
Could you also add a small note about the existence of this auth option in https://github.com/PrismarineJS/node-minecraft-protocol#echo-client-example ?
I think that lot of users will be asking about that and we want to make it as visible as possible

**Review by rom1504 — APPROVED** [review]

**Comment by rom1504** [comment]
LGTM, we probably want to support the json storage here too, but this can be done in another PR

**Comment by Kashalls** [comment]
> LGTM, we probably want to support the json storage here too, but this can be done in another PR

Afaik, we do not have enough information to support the `launcher_accounts.json` file. We definitely could do the same thing for Mojang Auth though.

**Comment by rom1504** [comment]
What information are we missing ? Can't we store the tokens ?

**Comment by craftycram** [comment]
is there a way to use an already valid access token to authenticate the user?

**Comment by kashalls** [comment]
> Is there a way to use an already valid access token to authenticate the user?
> 
> I have a web dashboard where I'm logging in using Microsoft Oauth with the XboxLive.signin scope. I want to use the access_token from there to create a minecraft client in the backend. That's why just using the already existing token, instead of using https://www.microsoft.com/link would be awesome.
> 
> EDIT: found your discord server and posted the question there as well. not sure if comments on three year old PR still get noticed

Afaik its been done, you just need to provide your own auth factory. Theres been some mentions in the discord on how to do this.

**Comment by craftycram** [comment]
As far as I understood browsing through the code doing the authentication flows normally I would need to fork prismarine-auth, since the flows are defined there right?  
There is no way for me to provide custom flows with minecraft-protocol client options directly, if I got that correctly.
@kashalls

**Comment by kashalls** [comment]
> As far as I understood browsing through the code doing the authentication flows normally I would need to fork prismarine-auth, since the flows are defined there right? There is no way for me to provide custom flows with minecraft-protocol client options directly, if I got that correctly. @kashalls

https://github.com/PrismarineJS/prismarine-auth/blob/master/src/MicrosoftAuthFlow.js#L54

The auth flow allows you to use a function instead of a caching location, allowing you to use external means of authentication.
Ex: https://discord.com/channels/413438066984747026/1127636567914184826

You can just replicate the data that prismarine-auth generates and store it like that. You can also try switching the auth provider to the old Yggdrasil one, setting all of the session data required and then setting the bot to offline mode. I haven't tried that in a bit however. 

> I have a web dashboard where I'm logging in using Microsoft Oauth with the XboxLive.signin scope. I want to use the access_token from there to create a minecraft client in the backend. That's why just using the already existing token

I would honestly save that token into a database that can be fetched using a cacheFactory like I posted a link of an example. It would make it easier.

**Comment by craftycram** [comment]
Aight thanks. Imma look into that later

**Pull request merged**